pkg/client/events_query/client.go

```diff
@@ -0,0 +1,273 @@
+	eqc.close()
+}
+
+// close unsubscribes all observers from all event subscription observables.
+func (eqc *eventsQueryClient) close() {
+	eqc.eventsBytesAndConnsMu.Lock()
+	defer eqc.eventsBytesAndConnsMu.Unlock()
+
+	for query, obsvblConn := range eqc.eventsBytesAndConns {
+		_ = obsvblConn.conn.Close()
+		obsvblConn.eventsBytes.UnsubscribeAll()
+
+		// remove closed eventsBytesAndConns
+		delete(eqc.eventsBytesAndConns, query)
+	}
+}
+
+// getNextRequestId increments and returns the JSON-RPC request ID which should
+// be used for the next request. These IDs are expected to be unique (per request).
+func (eqc *eventsQueryClient) getNextRequestId() string {
+	eqc.nextRequestId++
+	return fmt.Sprintf(requestIdFmt, eqc.nextRequestId)
+}
+
+// newEventwsBzAndConn creates a new eventsBytes and connection for the given query.
+func (eqc *eventsQueryClient) newEventsBytesAndConn(
+	ctx context.Context,
+	query string,
+) (*eventsBytesAndConn, error) {
+	conn, err := eqc.openEventsBytesAndConn(ctx, query)
+	if err != nil {
+		return nil, err
+	}
+
+	// Construct an eventsBytes for the given query.
+	eventsBzObservable, eventsBzPublishCh := channel.NewObservable[either.Either[[]byte]]()
+
+	// TODO_INVESTIGATE: does this require retry on error?
+	go eqc.goPublishEventsBz(ctx, conn, eventsBzPublishCh)
+
+	return &eventsBytesAndConn{
+		eventsBytes: eventsBzObservable,
+		conn:        conn,
+	}, nil
+}
+
+// openEventsBytesAndConn gets a connection using the configured dialer and sends
+// an event subscription request on it, returning the connection.
+func (eqc *eventsQueryClient) openEventsBytesAndConn(
+	ctx context.Context,
+	query string,
+) (client.Connection, error) {
+	// If no event subscription exists for the given query, create a new one.
+	// Generate a new unique request ID.
+	requestId := eqc.getNextRequestId()
+	req, err := eventSubscriptionRequest(requestId, query)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get a connection from the dialer.
+	conn, err := eqc.dialer.DialContext(ctx, eqc.cometWebsocketURL)
+	if err != nil {
+		return nil, ErrDial.Wrapf("%s", err)
+	}
+
+	// Send the event subscription request on the connection.
+	if err := conn.Send(req); err != nil {
+		subscribeErr := ErrSubscribe.Wrapf("%s", err)
+		// assume the connection is bad
+		closeErr := conn.Close()
+		return nil, multierr.Combine(subscribeErr, closeErr)
+	}
+	return conn, nil
+}
+
+// goPublishEventsBz blocks on reading messages from a websocket connection.
+// It is intended to be called from within a go routine.
+func (eqc *eventsQueryClient) goPublishEventsBz(
+	ctx context.Context,
+	conn client.Connection,
+	eventsBzPublishCh chan<- either.Either[[]byte],
+) {
+	// Read and handle messages from the websocket. This loop will exit when the
+	// websocket connection is closed and/or returns an error.
+	for {
+		event, err := conn.Receive()
+		if err != nil {
+			// TODO_CONSIDERATION: should we close the publish channel here too?
```

Comment:
I think we should add some logging / warning / printing here.

I believe it'll be particularly valuable in this stage of development.

Reply:
Where this will be used first, in the block client, there's a `WARN` log which covers this in a select/case that receives from an `errCh` in `retryOrError`. 

I also just fixed a bug that was preventing the async error in the `EventsBytes()` observable from propagating that direction, so now when this :point_up_2: branch is taken, that `WARN` will be logged.

Reply:
Yup!